OrbitService/ProducerSideServiceImpl.cpp

```diff
@@ -0,0 +1,324 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "ProducerSideServiceImpl.h"
+
+#include <thread>
+
+#include "OrbitBase/Logging.h"
+
+namespace orbit_service {
+
+void ProducerSideServiceImpl::OnCaptureStartRequested(CaptureEventBuffer* capture_event_buffer) {
+  CHECK(capture_event_buffer != nullptr);
+  LOG("About to send StartCaptureCommand to CaptureEventProducers (if any)");
+  {
+    absl::WriterMutexLock lock{&capture_event_buffer_mutex_};
+    capture_event_buffer_ = capture_event_buffer;
+  }
+  {
+    absl::MutexLock lock{&service_state_mutex_};
+    service_state_.capture_status = CaptureStatus::kCaptureStarted;
+  }
+}
+
+void ProducerSideServiceImpl::OnCaptureStopRequested() {
+  LOG("About to send StopCaptureCommand to CaptureEventProducers (if any)");
+  {
+    absl::MutexLock lock{&service_state_mutex_};
+    service_state_.capture_status = CaptureStatus::kCaptureStopping;
+
+    // Wait (for a limited amount of time) for all producers to send AllEventsSent or to disconnect.
+    service_state_mutex_.AwaitWithTimeout(absl::Condition(
+                                              +[](ServiceState* service_state) {
+                                                return service_state->producers_remaining == 0 ||
+                                                       service_state->exit_requested;
+                                              },
+                                              &service_state_),
+                                          max_wait_for_all_events_sent_);
+    CHECK(service_state_.producers_remaining >= 0);
```

Comment:
== 0 || exit requested?

Reply:
That would be false in case of timeout. I could do something like `!condition_was_satisfied || service_state_.producers_remaining == 0 || service_state->exit_requested` (I think? Or something like that), but that actually goes beyond the point of this `CHECK`, as it's basically restating the condition, while I just wanted to have a `CHECK` on `service_state_.producers_remaining`.

Reply:
maybe just remove the check in that case, I do not think it adds anything.

Reply:
well, I was under impression that producers_remaining cannot go negative. but ok. you can keep the check if you want to make sure it is not negative.